component/src/main/java/com/celements/photo/utilities/ImageUrlExtractor.java

```diff
@@ -0,0 +1,95 @@
+package com.celements.photo.utilities;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.TreeMap;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.xwiki.context.Execution;
+
+import com.celements.photo.container.ImageUrlDim;
+import com.xpn.xwiki.XWikiContext;
+import com.xpn.xwiki.web.Utils;
+
+public class ImageUrlExtractor {
+
+  public static final Logger LOGGER = LoggerFactory.getLogger(ImageUrlExtractor.class);
+
+  public static int MIN_SOCIAL_MEDIA_IMAGE_SIZE = 200;
+  public static int MIN_SOCIAL_MEDIA_AREA_SIZE = MIN_SOCIAL_MEDIA_IMAGE_SIZE
+      * MIN_SOCIAL_MEDIA_IMAGE_SIZE;
+
+  public List<ImageUrlDim> extractImagesList(String content) {
+    String regex = "<img .*?src=['\"](.*?)['\"].*?/>";
+    Map<Long, List<ImageUrlDim>> articleImages = new TreeMap<>();
+    Pattern p = Pattern.compile(regex);
+    Matcher m = p.matcher(content);
+    while (m.find()) {
+      String imgUrl = m.group(1);
+      Long key = getImgUrlSizeKey(imgUrl);
+      if (!articleImages.containsKey(key)) {
+        articleImages.put(key, new ArrayList<ImageUrlDim>());
+      }
+      articleImages.get(key).add(getImgUrlExternal(imgUrl));
+    }
+    return filterImagesByMissingOrSmallSize(articleImages);
+  }
+
+  List<ImageUrlDim> filterImagesByMissingOrSmallSize(Map<Long, List<ImageUrlDim>> articleImages) {
+    List<ImageUrlDim> sortedImages = new ArrayList<>();
+    for (Long imgArea : articleImages.keySet()) {
+      // Image size not extractable is in key == '-1' Don't include too small images.
+      if ((imgArea == -1) || (imgArea >= (MIN_SOCIAL_MEDIA_AREA_SIZE))) {
+        Collections.reverse(articleImages.get(imgArea));
+        sortedImages.addAll(articleImages.get(imgArea));
+      }
+    }
+    return sortedImages;
+  }
+
+  ImageUrlDim getImgUrlExternal(String imgUrl) {
+    if (!imgUrl.startsWith("http://") && !imgUrl.startsWith("https://")) {
+      String action = imgUrl.replaceAll("^.*?/(.*?)/.*$", "$1");
+      String space = imgUrl.replaceAll("^(.*?/){2}(.*?)/.*$", "$2");
+      String docname = imgUrl.replaceAll("^(.*?/){3}(.*?)/.*$", "$2");
+      String filename = imgUrl.replaceAll("^(.*?/){4}(.*?)(|\\?.*)$", "$2");
+      String query = imgUrl.replaceAll("^.*\\?(.*)$", "$1");
+      imgUrl = getContext().getURLFactory().createAttachmentURL(filename, space, docname, action,
+          query, getContext().getDatabase(), getContext()).toString();
+    }
+    return new ImageUrlDim(imgUrl, parseImgUrlDimension(imgUrl, "celwidth"), parseImgUrlDimension(
+        imgUrl, "celheight"));
+  }
+
+  Long getImgUrlSizeKey(String imgUrl) {
+    int w = parseImgUrlDimension(imgUrl, "celwidth");
+    int h = parseImgUrlDimension(imgUrl, "celheight");
+    Long area = new Long(h * w);
```

Comment:
The two int can't overflow, since Integer.parseInt() throws a NumberFormatException if it would overflow.
However, (max int)^2 can overflow a Long. I doubt an image with more than 2^63-1 pixels will be practical on a web page anytime in the foreseeable future. Nevertheless I restricted width and hight to max 10^9 each.